File: src/main/java/org/saigon/striker/service/AgodaService.java
package org.saigon.striker.service;

import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.ClientResponse;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.regex.Pattern;

import static java.lang.String.format;

@Service
public class AgodaService {

    public static final String AGODA_BASE_URL = "https://www.agoda.com";
    public static final String AGODA_COOKIE_NAME = "agoda.version.03";

    private static final Pattern COOKIE_ID_PATTERN = Pattern.compile(
            "CookieId=([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})");

    private final WebClient webClient;

    public AgodaService(WebClient.Builder webClientBuilder) {
        this.webClient = webClientBuilder.baseUrl(AGODA_BASE_URL).build();
    }

    public Mono<String> getAgodaCookie() {
        // TODO error handling
        return webClient.get().accept(MediaType.TEXT_HTML)
                .exchange()
                .flatMap(clientResponse -> clientResponse.toEntity(String.class))
                .map(responseEntity -> getCookieId(responseEntity.getHeaders().get(HttpHeaders.SET_COOKIE)));
    }

    public Mono<ClientResponse> callAgodaApi(String uri, String currency) {
        // TODO error handling
        return getAgodaCookie().flatMap(cookieId ->
                webClient.get()
                        .uri(uri).accept(MediaType.APPLICATION_JSON).cookie(AGODA_COOKIE_NAME, createCookie(cookieId, currency))
                        .exchange());
    }

    public static String createCookie(String cookieId, String currency) {
        return format("CookieId=%s&CurLabel=%s", cookieId, currency);
    }

    private static String getCookieId(List<String> cookies) {
        if (cookies == null) {
            throw new IllegalStateException("Cookie headers are missing");
        }
        return cookies.stream()
                .filter(cookie -> cookie.contains(AGODA_COOKIE_NAME))
                .map(AgodaService::parseCookieId)
                .findFirst().orElseThrow(() -> new IllegalStateException(format("Cookie '%s' is missing", AGODA_COOKIE_NAME)));
    }

    private static String parseCookieId(String cookie) {
        var matcher = COOKIE_ID_PATTERN.matcher(cookie);
        if (matcher.find()) {
            return matcher.group(1);
        } else {
            throw new IllegalStateException(format("Cookie '%s' is missing CookieId", AGODA_COOKIE_NAME));
        }
    }
}

File: src/main/kotlin/org/saigon/striker/service/AgodaService.kt
package org.saigon.striker.service

import org.saigon.striker.model.*
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.http.MediaType.TEXT_HTML
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.client.awaitEntity
import org.springframework.web.reactive.function.client.awaitExchange
import org.springframework.web.util.UriComponentsBuilder

@Component
class AgodaService(webClientBuilder: WebClient.Builder, baseUrl: String = "https://www.agoda.com") {

    private val cookieName = "agoda.version.03"
    private val cookieIdRegex = Regex("CookieId=([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})")

    val webClient = webClientBuilder.baseUrl(baseUrl).build()

    suspend fun getHotel(parameters: AgodaParameters): Hotel {
        val agodaHotel = webClient.get()
            .uri(createHotelUri(parameters))
            .accept(APPLICATION_JSON)
            .cookie(cookieName, createCookie(getCookieId(), parameters.currency))
            .retrieve()
            .onStatus(HttpStatus::isError) { Exceptions.handleAgodaError(it) }
            .awaitBody<AgodaHotel>()

        return agodaHotel.toHotel()
    }

    suspend fun search(query: String): SearchResult {
        val agodaSearchResult = webClient.get()
            .uri(createSearchUri(query))
            .accept(APPLICATION_JSON)
            .retrieve()
            .onStatus(HttpStatus::isError) { Exceptions.handleAgodaError(it) }
            .awaitBody<AgodaSearchResult>()

        return agodaSearchResult.toSearchResult()
    }

    suspend fun getCookieId(): String {
        val responseEntity = webClient.get()
            .uri("/")
            .accept(TEXT_HTML)
            .awaitExchange()
            .awaitEntity<String>() // retrieve body to release resources

        if (responseEntity.statusCode.isError) {
            throw AgodaApiException(responseEntity.statusCode)
        }
        return getCookieId(responseEntity.headers[HttpHeaders.SET_COOKIE])
    }

    private fun getCookieId(cookies: List<String>?): String {
        val cookieValue = cookies?.find { it.contains(cookieName) }
            ?: throw IllegalStateException("Agoda cookie is missing")

        return cookieIdRegex.find(cookieValue)?.groupValues?.get(1)
            ?: throw IllegalStateException("Agoda cookie id is missing")
    }

    private fun createCookie(cookieId: String, currency: String): String {
        return "CookieId=$cookieId&CurLabel=$currency"
    }

    private fun createHotelUri(parameters: AgodaParameters): String {
        return UriComponentsBuilder.fromPath("/api/en-us/pageparams/property")
            .queryParam("hotel_id", parameters.hotelId)
            .queryParam("checkIn", parameters.checkInDate)
            .queryParam("los", parameters.lengthOfStay)
            .queryParam("rooms", parameters.rooms)
            .queryParam("adults", parameters.adults)
            .queryParam("childs", parameters.children)
            .toUriString()
    }

    private fun createSearchUri(query: String): String {
        return UriComponentsBuilder.fromPath("/Search/Search/GetUnifiedSuggestResult/3/1/1/0/en-us")
            .queryParam("searchText", query)
            .queryParam("isHotelLandSearch", "true")
            .toUriString()
    }
}
